TensorFlow Lite filter - link issue depending on delegates supported by TFlite lib

Hello all,

# Issue Description
A change has been merged recently into main branch changing the configuration options for NNStreamer TFLite filter:
https://github.com/nnstreamer/nnstreamer/commit/2a86a9c9147a8b59693af9dbee4b96876fb035f9
Before this change, delegates support of GPU, NNAPI and XNNPACK was configurable via meson option.
It is no longer the case and from now on, support of this 3 delegates is expected by NNStreamer filter to be always available.

This is not an issue at build time because headers for every delegate are exported by the lib so can be #include-d by filter.
However it is an issue at  link time, because delegates support in TFLite lib is defined at library configuration time by its cmake:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/CMakeLists.txt#L64
It still stands true on main branch of tensorflow upstream project: it can not be expected that all delegates are built in the lib.
When one delegate is not compiled in TFLite lib, NNStreamer TFLite filter can not link, because it references symbols missing in TFLite lib.  

We have the issue using a TFLite 2.6.0 library without GPU delegate compiled in - error output below.
But similar issue would presumably show up with any missing delegate and with any TFLite version.

Error at link reported during subplugin (.so) build:
tensor_filter_tensorflow_lite.cc.o: in function `TFLiteInterpreter::loadModel(int, tflite_delegate_e)::{lambda(TfLiteDelegate*)#2}::_FUN(TfLiteDelegate*)':
tensor_filter_tensorflow_lite.cc:475: undefined reference to `TfLiteGpuDelegateV2Delete'
tensor_filter_tensor_filter_tensorflow_lite.cc.o: in function `TFLiteInterpreter::loadModel(int, tflite_delegate_e)':
tensor_filter_tensorflow_lite.cc:458: undefined reference to `TfLiteGpuDelegateOptionsV2Default'
tensor_filter_tensorflow_lite.cc:472: undefined reference to `TfLiteGpuDelegateV2Create'
collect2: error: ld returned 1 exit status
 

Expected Result
============
NNStreamer tflite filter should be able to link with multiple flavours of TFLite library, even when all delegates (XNNPACK, GPU, NNAPI) are not built-in.
Note: same NNStreamer change has added a TODO to remove the meson option for external delegate. Doing so would presumably lead to the same issue.

Can the previous meson options be reenabled, as well as conditional build in the NNStreamer tflite filter of the delegate-specific block
option('tflite2-gpu-delegate-support', type: 'boolean', value: 'false')
option('tflite2-nnapi-delegate-support', type: 'boolean', value: 'false')
option('tflite2-xnnpack-delegate-support', type: 'boolean', value: 'false')


How to Reproduce
===============
Build NNStreamer TFLite filter on a file system hosting a TFLite library without GPU delegate.


Thank you for your support.

--- a/ext/nnstreamer/tensor_filter/meson.build
+++ b/ext/nnstreamer/tensor_filter/meson.build
@@ -223,9 +223,40 @@ if tflite2_support_is_available
     tflite2_compile_args += '-DUSE_TENSORFLOW2_HEADER_PATH=1'
   endif
 
-  # TODO: tflite v2.8.0+ supports build option for external delegate
-  # remove below option after that.
-  if get_option('tflite2-external-delegate-support')
+  # Check each delegate linking with given tflite library
+  ## xnnpack delegate
+  if cxx.links('''
+        #include "tensorflow/lite/delegates/xnnpack/xnnpack_delegate.h"
+        TfLiteXNNPackDelegateOptions xnnpack_options = TfLiteXNNPackDelegateOptionsDefault ();
+        int main() {return 0;}
+      ''', dependencies : [tflite2_support_deps, thread_dep], name : 'xnnpack delegate')
+    tflite2_compile_args += '-DTFLITE_XNNPACK_DELEGATE_SUPPORTED'
+  endif
+
+  ## gpu delegate
+  if cxx.links('''
+        #include "tensorflow/lite/delegates/gpu/delegate.h"
+        TfLiteGpuDelegateOptionsV2 gpu_options = TfLiteGpuDelegateOptionsV2Default ();
+        int main() {return 0;}
+      ''', dependencies : [tflite2_support_deps, libdl_dep], name : 'gpu delegate')
+    tflite2_compile_args += '-DTFLITE_GPU_DELEGATE_SUPPORTED'
+  endif
+
+  ## nnapi delegate
+  if cxx.links('''
+        #include "tensorflow/lite/delegates/nnapi/nnapi_delegate.h"
+        tflite::StatefulNnApiDelegate::StatefulNnApiDelegate ();
+        int main() {return 0;}
+      ''', dependencies : [tflite2_support_deps], name : 'nnapi delegate')
+    tflite2_compile_args += '-DTFLITE_NNAPI_DELEGATE_SUPPORTED'
+  endif
+
+  ## external delegate
+  if cxx.links('''
+      #include "tensorflow/lite/delegates/external/external_delegate.h"
+      TfLiteExternalDelegateOptions external_options = TfLiteExternalDelegateOptionsDefault ("");
+      int main() {return 0;}
+    ''', dependencies : [tflite2_support_deps, libdl_dep], name : 'external delegate')
     tflite2_compile_args += '-DTFLITE_EXTERNAL_DELEGATE_SUPPORTED'
   endif
 

--- a/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc
+++ b/ext/nnstreamer/tensor_filter/tensor_filter_tensorflow_lite.cc
@@ -52,14 +52,27 @@
 #  include <tensorflow/contrib/lite/model.h>
 #endif
 
-#if TFLITE_VERSION_MAJOR >= 2
+/** control delegate headers */
+#ifdef TFLITE_XNNPACK_DELEGATE_SUPPORTED
 #  if USE_TENSORFLOW2_HEADER_PATH
 #    include <tensorflow2/lite/delegates/xnnpack/xnnpack_delegate.h>
-#    include <tensorflow2/lite/delegates/gpu/delegate.h>
-#    include <tensorflow2/lite/delegates/nnapi/nnapi_delegate.h>
 #  else
 #    include <tensorflow/lite/delegates/xnnpack/xnnpack_delegate.h>
+#  endif
+#endif
+
+#ifdef TFLITE_GPU_DELEGATE_SUPPORTED
+#  if USE_TENSORFLOW2_HEADER_PATH
+#    include <tensorflow2/lite/delegates/gpu/delegate.h>
+#  else
 #    include <tensorflow/lite/delegates/gpu/delegate.h>
+#  endif
+#endif
+
+#ifdef TFLITE_NNAPI_DELEGATE_SUPPORTED
+#  if USE_TENSORFLOW2_HEADER_PATH
+#    include <tensorflow2/lite/delegates/nnapi/nnapi_delegate.h>
+#  else
 #    include <tensorflow/lite/delegates/nnapi/nnapi_delegate.h>
 #  endif
 #endif
@@ -428,7 +441,7 @@ TFLiteInterpreter::loadModel (int num_threads, tflite_delegate_e delegate_e)
   switch (delegate_e) {
     case TFLITE_DELEGATE_XNNPACK:
     {
-#if TFLITE_VERSION_MAJOR >= 2
+#if TFLITE_XNNPACK_DELEGATE_SUPPORTED
       /* set xnnpack delegate */
       TfLiteXNNPackDelegateOptions xnnpack_options =
           TfLiteXNNPackDelegateOptionsDefault();
@@ -452,7 +465,7 @@ TFLiteInterpreter::loadModel (int num_threads, tflite_delegate_e delegate_e)
     }
     case TFLITE_DELEGATE_GPU:
     {
-#if TFLITE_VERSION_MAJOR >= 2
+#if TFLITE_GPU_DELEGATE_SUPPORTED
       /* set gpu delegate when accelerator set to GPU */
       TfLiteGpuDelegateOptionsV2 options = TfLiteGpuDelegateOptionsV2Default ();
       options.experimental_flags = TFLITE_GPU_EXPERIMENTAL_FLAGS_NONE;
@@ -482,7 +495,7 @@ TFLiteInterpreter::loadModel (int num_threads, tflite_delegate_e delegate_e)
     }
     case TFLITE_DELEGATE_NNAPI:
     {
-#if TFLITE_VERSION_MAJOR >= 2
+#if TFLITE_NNAPI_DELEGATE_SUPPORTED
       /* set nnapi delegate when accelerator set to auto (cpu.neon in Android) or NPU */
       delegate = new tflite::StatefulNnApiDelegate ();
       void (* deleter) (TfLiteDelegate *) =

--- a/meson_options.txt
+++ b/meson_options.txt
@@ -27,7 +27,6 @@ option('trix-engine-support', type: 'feature', value: 'auto')
 option('enable-test', type: 'boolean', value: true)
 option('install-test', type: 'boolean', value: false)
 option('enable-pytorch-use-gpu', type: 'boolean', value: false) # default value, can be specified at run time
-option('tflite2-external-delegate-support', type: 'boolean', value: 'false')
 option('enable-mediapipe', type: 'boolean', value: false)
 option('enable-env-var', type: 'boolean', value: true)
 option('enable-symbolic-link', type: 'boolean', value: true)

